internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,202 @@
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+// resetBuffer drops all buffered logs (in the case of a successful send, or an
+// unrecoverable error).
+func (c *Client) resetBuffer() {
+	c.entries = c.entries[:0]
+}
+
+// truncateBuffer truncates the buffer to maxRequestEntries, removing the
+// oldest logs first.
+func (c *Client) truncateBuffer() {
+	l := len(c.entries)
+	if l > maxRequestEntries {
+		c.entries = c.entries[(l - maxRequestEntries):]
+	}
+}
+
+func (c *Client) buildRequest() lokiRequest {
+	// Put entries into service "buckets"
+	bucketedEntries := map[string][]lokiEntry{}
+	for _, data := range c.entries {
+		bucketedEntries[data.service] = append(bucketedEntries[data.service], data.entry)
+	}
+
+	var req lokiRequest
+	for service, entries := range bucketedEntries {
+		stream := lokiStream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+type lokiRequest struct {
+	Streams []lokiStream `json:"streams"`
+}
+
+type lokiStream struct {
+	Labels  map[string]string `json:"stream"`
+	Entries []lokiEntry       `json:"values"`
+}
+
+type lokiEntry [2]string
+
+type lokiEntryWithService struct {
+	entry   lokiEntry
+	service string
+}
+
+// handleServerResponse determines what to do based on the response from the
+// Loki server. 4xx and 5xx responses indicate errors, so in this case, we will
+// bubble up the error to the caller.
+func (c *Client) handleServerResponse(resp *http.Response) error {
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	code := resp.StatusCode
+	switch {
+	case code == 200 || code == 204:
+		// Success - safe to drop logs
+		c.resetBuffer()
+		return nil
+
+	case code == http.StatusTooManyRequests:
+		// For 429, don't drop logs - just retry later
+		return errFromResponse(resp)
+
+	case 400 <= code && code < 500:
+		// 4xx indicates a client problem, so drop the logs (retrying won't help)
+		logger.Noticef("Target %q: request failed with status %d, dropping %d logs",
+			c.targetName, code, c.NumBuffered())
+		c.resetBuffer()
+		return errFromResponse(resp)
+
+	case 500 <= code && code < 600:
+		// 5xx indicates a problem with the server, so don't drop logs (retry later)
+		return errFromResponse(resp)
+
+	default:
+		// Unexpected response - don't drop logs to be safe
+		return fmt.Errorf("unexpected response from server: %v", resp.Status)
+	}
+}
+
+// errFromResponse generates an error from a failed *http.Response.
+// NB: this function reads the response body.
```

Comment:
I've forgotten my Latin and what NB means. Perhaps just:
```suggestion
// This function reads the response body.
```

Reply:
"NB:" == "Note:"

Reply:
sorry thought that was well known